RFC-0004-sparse-fill-value/README.md

```diff
@@ -0,0 +1,509 @@
+       all functions supporting sparse inputs, we may need a more
+       efficient way than `.nonzero().numel() == 0` to determine if a
+       tensor contains only zeros. This efficiency issue is relevant
+       only for hybrid tensors with a large dense part.
+
+    2. Update the related functions to handle non-zero fill-values of
+       input sparse tensors correctly.
+
+       For instance, consider a matrix multiplication of two sparse
+       tensors `A` and `B` with fill-values `a` and `b`, respectively,
+       then the `matmul` operation can be expanded as follows:
+
+       ```python
+       matmul(A, B) = matmul(A - fA + fA, B - fB + fB)
+                    = matmul(A - fA, B - fB) + fA * matmul(ones_like(A), B) + fB * matmul(A, ones_like(B))
+       ```
+
+       where the first term can be computed using existing matmul for
+       sparse tensors with zero fill-value, and the last two terms can
+       be replaced with a computation of a single row or column of the
+       corresponding matrix products that has reduced computational
+       complexity.
+
+10. Sparse tensors with defined fill-value have intrinsic constraints
+    between all the unspecified tensor elements (these are always
+    equal) that must be taken into account when implementing Autograd
+    backward methods for functions that receive sparse tensors as
+    inputs.
+
+    Sparse tensors with undefined fill-value don't have the intrinsic
+    constraints as discussed above.
+
+
+### Future extensions and existing issues
+
+11. For Graph domain, the undefined fill-value can be specified as a
+    tensor with zero dimension(s) that satisfies all the relations
+    listed above except point 5. Invalidation of the point 5 will
+    provide a consistent way to differentiate between defined and
+    undefined fill-values.
+
+    For instance, to reset a fill-value of a sparse tensor to undefined
+    fill-value, one can use:
+
+    ```python
+    A.fill_value().resize_((0,) * len(A.values().shape[1:]))
+    ```
+
+    Note that this operation is reversible, that is, to reset a
+    undefined fill-value to a defined value, say `1.2`, one can use:
+
+    ```python
+    A.fill_value().resize_(A.values().shape[1:]).fill_(1.2)
+    ```
+
+    This works only when the initial fill-value of `A` is a defined
+    value with the shape constraints specified in point 5.
+
+12. For Statistics domain, where unspecified elements are interpreted
+    as "missing data". The missing data concept is different from
+    undefined value in the sense that it represents a certain
+    measurement flaw that cannot be foreseen to appear in a data while
+    the undefined value represents a structural lack of data that is
+    determined by the system under study.
+
+    The fill-value corresponding to missing data can be specified as
+    `nan` (or a tensor of `nan`-s in case of hybrid tensor). This
+    choice is consistent with convention used in Python Pandas
+    package.
```

Comment:
This choice is problematic, and Pandas is introducing its own `NA` support instead of using `nan`. The reason is that it doesn't work for integers, because there is no `nan`.

If fill value was scalar, this would be much easier, use an empty tensor for example.

Reply:
I agree, `nan` is a bad choice, in general. PyTorch currently supports only float sparse tensors so the lack-of-nan-for-an-integer does not apply, at least, not currently.

The shape of fill-value (scalar or no scalar) is irrelevant, empty tensors can have arbitrary shapes provided that at least one dimension size is 0.

That said, I now think that I got the usage of sparse tensors in Statistics domain wrong. "missing data" exist as specified values (be them `nan`, `N/A`, etc), the fill-value represents elements havng the most common value. I'll fix that..

Reply:
A reasonable fix seems to be removing the discussion of Statistics domain from the document.

Reply:
Removed extension to Statistics domain as not needed.